Userland/xargs.cpp

```diff
@@ -0,0 +1,285 @@
+
+    if (!split_with_nulls && strlen(specified_delimiter) > 1) {
+        fprintf(stderr, "xargs: the delimiter must be a single byte\n");
+        return 1;
+    }
+
+    char entry_separator = split_with_nulls ? '\0' : *specified_delimiter;
+
+    StringView placeholder_view { placeholder };
+
+    if (arguments.is_empty())
+        arguments.append("echo");
+
+    Vector<Vector<StringView>> argument_parts;
+    argument_parts.ensure_capacity(arguments.size());
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        size_t previous_index = 0;
+        if (!placeholder_view.is_empty()) {
+            if (auto first_index = arg.find_first_of(placeholder_view); first_index.has_value()) {
+                some_argument_has_replacement = true;
+                while (first_index.has_value()) {
+                    auto current_index = first_index.value();
+                    auto substring_until_replacement = arg.substring_view(previous_index, current_index);
+                    parts.append(substring_until_replacement);
+                    previous_index = current_index;
+                    first_index = arg.substring_view(current_index + placeholder_view.length(), arg.length() - current_index - placeholder_view.length()).find_first_of(placeholder_view);
+                }
+                parts.append(arg.substring_view(previous_index + placeholder_view.length(), arg.length() - previous_index - placeholder_view.length()));
+            } else {
+                parts.append(arg);
+            }
+        } else {
+            parts.append(arg);
+        }
+        argument_parts.append(move(parts));
+    }
+
+    // Append a '{}' at the end if no arguments have replacements.
+    if (!some_argument_has_replacement) {
+        Vector<StringView> parts;
+        parts.append("");
+        parts.append("");
+        argument_parts.append(move(parts));
+    }
+
+    FILE* fp = stdin;
+    bool is_stdin = true;
+
+    if (StringView { "-" } != file_to_read) {
+        // A file was specified, try to open it.
+        fp = fopen(file_to_read, "r");
+        if (!fp) {
+            perror("fopen");
+            return 1;
+        }
+        is_stdin = false;
+    }
+
+    StringBuilder builder;
+    Vector<char*> child_argv;
+
+    int devnull_fd = 0;
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    size_t total_command_length = 0;
+    size_t lines_used_for_this_command = 0;
+
+    auto fail = read_lines(fp, entry_separator, [&](StringView line) {
+        if (line.ends_with('\n'))
+            line = line.substring_view(0, line.length() - 1);
+
+        if (line.is_empty())
+            return Continue;
+
+        if (lines_used_for_this_command == 0) {
```

Comment:
My understanding is that we process the first line differently for two reasons:
* It must fit into `max_lines` (makes no sense otherwise)
* We substitute the first line in place of the placeholder; whereas we simply append other lines as separate args

Might be a good idea to spell this out in a comment.

Also perhaps rename `line`s to `item`s everywhere?